### PR TITLE
Progress Dialog: Fix race on PPU compilation status

### PR DESCRIPTION
**Pull request by elad335**
If more than one PPU thread was loading PPU PRX/OVL, there could have been a race that would make the "Compiling PPU Modules" message never disappear because one thread was storing the string the other thread stored. (g_progr never gets null again)

### CODE REVIEW

File: rpcs3/Emu/Cell/PPUThread.cpp [hunk 1]
```diff
@@ -4523,10 +4523,13 @@ bool ppu_initialize(const ppu_module& info, bool check_only)
 		// Check object file
 		if (jit_compiler::check(cache_path + obj_name))
 		{
-			if (!jit && !check_only)
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/PPUThread.cpp:4526]: Why was this changed?

Comment (elad335) [reply rpcs3/Emu/Cell/PPUThread.cpp:4526]: It was a bug, g_progr_ptotal is incremented if check_only is false. It does not care about null jit.

File: rpcs3/Emu/system_progress.hpp [hunk 2]
```diff
@@ -15,21 +27,52 @@ extern atomic_t<bool> g_system_progress_stopping;
...
+			progr.m_updates++;
+		});
+
+		return *this;
+	}
+
 	~scoped_progress_dialog() noexcept
 	{
-		g_progr.release(m_prev);
+		g_progr.atomic_op([this](progress_dialog_string_t& progr)
```
Comment (Megamouse) [inline rpcs3/Emu/system_progress.hpp:63]: I don't see the purpose in this.
You're supposed to reset g_progr on scope exit.
It's not supposed to be updated by 2 different entities simultaneously.

Comment (elad335) [reply rpcs3/Emu/system_progress.hpp:66]: Well, it kind of is updated by 2 different PPUs if they call the same syscalls sys_prx or sys_overlay syscalls.
The perfect mechanism would be keeping a thread-safe array of strings but it seems overkill for this purpose.

File: rpcs3/Emu/system_progress.hpp [hunk 3]
```diff
@@ -15,21 +27,55 @@ extern atomic_t<bool> g_system_progress_stopping;
...
 	// Saved previous value
-	const char* const m_prev;
+	const char* m_prev;
+	u32 m_prev_update;
 
 public:
 	scoped_progress_dialog(const char* text) noexcept
-		: m_prev(g_progr.exchange(text ? text : ""))
 	{
+		std::tie(m_prev, m_prev_update) = g_progr.atomic_op([text = text ? text : ""](progress_dialog_string_t& progr)
```
Comment (Megamouse) [inline rpcs3/Emu/system_progress.hpp:36]: Why `text = text ? text : ""` ?

File: rpcs3/Emu/system_progress.hpp [hunk 4]
```diff
@@ -15,21 +27,55 @@ extern atomic_t<bool> g_system_progress_stopping;
...
+		});
 	}
 
 	scoped_progress_dialog(const scoped_progress_dialog&) = delete;
 
 	scoped_progress_dialog& operator=(const scoped_progress_dialog&) = delete;
 
+	scoped_progress_dialog& operator=(const char* text) noexcept
+	{
+		m_prev_update = g_progr.atomic_op([this, text = text ? text : ""](progress_dialog_string_t& progr)
```
Comment (Megamouse) [inline rpcs3/Emu/system_progress.hpp:52]: Why text = text ? text : "" ?

File: rpcs3/Emu/system_progress.hpp [hunk 5]
```diff
@@ -15,21 +27,55 @@ extern atomic_t<bool> g_system_progress_stopping;
...
+
 	~scoped_progress_dialog() noexcept
 	{
-		g_progr.release(m_prev);
+		g_progr.atomic_op([this](progress_dialog_string_t& progr)
+		{
+			if (progr.m_user_count-- == 1)
+			{
+				// Clean text only on last user
+				progr.m_text = nullptr;
```
Comment (Megamouse) [inline rpcs3/Emu/system_progress.hpp:71]: How do you know this was the state before?
That's pure speculation.

File: rpcs3/Emu/system_progress.hpp [hunk 6]
```diff
@@ -15,21 +27,55 @@ extern atomic_t<bool> g_system_progress_stopping;
...
+			}
+			else if (m_prev && m_prev_update == progr.m_updates)
```
Comment (Megamouse) [inline rpcs3/Emu/system_progress.hpp:73]: why should this only happen if m_prev is not null ?

Comment (elad335) [reply rpcs3/Emu/system_progress.hpp:73]: This is a wrong condition you are right, it must never be null if there are active users.

File: rpcs3/Emu/system_progress.hpp (hunk 5 above)
Comment (elad335) [reply rpcs3/Emu/system_progress.hpp:71]: Added assert.

File: rpcs3/Emu/system_progress.hpp [hunk 7]
```diff
@@ -15,21 +27,57 @@ extern atomic_t<bool> g_system_progress_stopping;
 class scoped_progress_dialog final
 {
 	// Saved previous value
-	const char* const m_prev;
+	const char* m_prev;
+	u32 m_prev_update;
```
Comment (Megamouse) [inline rpcs3/Emu/system_progress.hpp:31]: Just use a static ID counter. Much easier to understand than some "update counter"

Comment (Megamouse) [reply rpcs3/Emu/system_progress.hpp:31]: Like `m_id = counter++;` in constructor.
And simply set g_progr.last_user_id on update.

**Comment by Megamouse** [comment]
lgtm. But I haven't tested anything.

**Comment by Megamouse** [comment]
Sadly it's broken AF xD

File: rpcs3/Emu/Cell/PPUThread.cpp (hunk 1 above)
Comment (Megamouse) [reply rpcs3/Emu/Cell/PPUThread.cpp:4526]: Actually, this change is wrong.
It's incremented later again if it needs to be compiled a couple of lines below.

So you're incrementing the uncached ones twice now.

File: rpcs3/Emu/system_progress.hpp (hunk 7 above)
Comment (elad335) [reply rpcs3/Emu/system_progress.hpp]: Actuallly this merhod does not wirk on a large scale with 3 or more users.

Comment (elad335) [reply rpcs3/Emu/system_progress.hpp]: Actually this method was flawed, and using the previous method with a tweak allows it it to be perfect.
